File: src/main/java/tech/bobliu/assignment07/service/MessageService.java
package tech.bobliu.assignment07.service;

import tech.bobliu.assignment07.model.Message;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

public class MessageService {
    private static final List<Message> messages = new ArrayList<>();

    public static List<Message> getMessagesTo(int targetId, Integer afterId) {
        int start = (afterId == null ? -1 : afterId) + 1;
        return messages.stream()
                .skip(start)
                .filter(m -> m.getTargetId() == targetId || m.getSenderId() == targetId || m.getTargetId() == -1)
                .collect(Collectors.toList());
    }

    public static int sendMessage(String content, int senderId, int targetId) {
        synchronized (messages) {
            Message message = new Message(messages.size(), senderId, targetId, content, new Date());
            messages.add(message);
            return message.getId();
        }
    }
}

File: src/main/kotlin/tech/bobliu/assignment07/service/MessageService.kt
package tech.bobliu.assignment07.service

import tech.bobliu.assignment07.model.Message

object MessageService {
    private val messages: ArrayList<Message> = ArrayList()

    fun getMessagesTo(targetId: Int, afterId: Int?) = messages
        .drop((afterId ?: -1) + 1).filter { it.targetId == targetId || it.senderId == targetId || it.targetId == -1 }

    fun sendMessage(content: String, senderId: Int, targetId: Int): Int {
        synchronized(messages) {
            val message = Message(
                id = messages.size,
                senderId = senderId,
                targetId = targetId,
                content = content,
            )
            messages.add(message)
            return message.id
        }
    }
}